README.md
Custom HTTP client for the UMMobile app

# Getting Started
## Initialization
A base URL and a static authorization can be defined in the initialization.
```dart
// Simple initalization
final http = UMMobileCustomHttp(baseUrl: 'https://jsonplaceholder.typicode.com');

// Initialiation with authentication
final http = UMMobileCustomHttp(
  baseUrl: 'https://jsonplaceholder.typicode.com',
  auth: Auth(
    token: () => 'YOUR_TOKEN',
    tokenType: 'Bearer',
    headername: 'Authorization',
  ),
);
```

## HTTP Calls
To make a call you need to use the function that start with `custom`.
```dart
final http = UMMobileCustomHttp(baseUrl: 'https://jsonplaceholder.typicode.com');\

await http.customGet(path: '/posts/1');
await http.customPost(path: '/posts/1', body: {});
await http.customDelete(path: '/posts/1');
```

You can also pass a mapper function to format the response body and return a specific type.
```dart
final http = UMMobileCustomHttp(baseUrl: 'https://jsonplaceholder.typicode.com');

// Return a Post instance
await http.customGet<Post>(
  path: '/posts/1',
  mapper: (json) => Post(
    userId: json['userId'],
    id: json['id'],
    title: json['title'],
    body: json['body'],
  ),
);
```

## Exceptions
Throws an `HttpCallException` if an error occurs. The `HttpCallException` contains a `type` attribute that can be as shown in the list below:

- `ClientError`: when the status code is between 300 and 499.
- `ServerError`: when the status code is between 500 and 599.
- `ConnectionError`: when a connection error occurs and cannot be specified.
- `ServerDown`: when cannot connect to the backend.
- `ClientOffline`: when cannot connect to `yahoo.com`.
- `ExpiredToken`: when access token is expired & need to be refresh.
- `Unauthorized`: when an error occurred with the authorization.
- `Other`: when any other `Exception` occurs.

You can use the Exception like:
```dart
final http = UMMobileCustomHttp(baseUrl: 'https://jsonplaceholder.typicode.com');
try {
  await http.customGet(path: '/posts/1');
} on ConnectionErrorException catch(e) {
  if(e.type === HttpException.ClientOffline) {
    // display that the client have no connection
  }
} on ClientErrorException catch(e) {
  if(e.type === HttpException.ExpiredToken) {
    // display that the token is expired
  }
}
```


CHANGELOG.md
## 0.0.1
> Release date: 07/Oct/2021
- First version

## 0.0.2-beta1
> Release date: 08/Oct/2021
- Adds checkInternetError function
- Adds more HttpExceptions
- Updates docs

## 0.0.2-beta2
> Release date: 11/Oct/2021
- Updates `utf8Decode` default value to `false`

## 0.0.2
> Release date: 20/Oct/2021
- Adds `queries` argument

## 0.0.3-beta1
> Release date: 26/Oct/2021
- Adds ExpiredToken exception
- Adds validation for JWT
- Improves Auth class usage

## 0.0.3-beta2
> Release date: 28/Oct/2021
- Fixes format of authorization header value
- Adds Unauthorized exception
- Updates HttpCallExceptions implementations:
  - Adds ClientErrorException
  - Adds ServerErrorException
  - Adds ConnectionErrorException
- Refactor checkInternetConnection:
  - Renames to throwConnectionException
  - Updates to throw a ConnectionErrorException instead of returning HttpException type

## 0.0.3
> Release date: 01/Nov/2021
- Adds ExpiredToken exception
- Adds validation for JWT
- Improves Auth class usage
- Fixes format of authorization header value
- Adds Unauthorized exception
- Updates HttpCallExceptions implementations:
  - Adds ClientErrorException
  - Adds ServerErrorException
  - Adds ConnectionErrorException
- Refactor checkInternetConnection:
  - Renames to throwConnectionException
  - Updates to throw a ConnectionErrorException instead of returning HttpException type


lib/src/exceptions/http_call_exception.dart
import 'package:get/get_connect.dart';
import 'package:ummobile_custom_http/src/enums/http_exceptions.dart';

export 'client_error_exception.dart';
export 'connection_error_exception.dart';
export 'server_error_exception.dart';

// The Exception that is thrown if an error ocurrs.
class HttpCallException implements Exception {
  /// The type of Exception.
  ///
  /// Can be as shown in the table below:
  ///
  /// |Type                      | Case                                                   |
  /// |--------------------------|--------------------------------------------------------|
  /// | `ClientError`            | When the status code is between 300 and 499            |
  /// | `ExpiredToken`           | When access token is expired & need to be refresh      |
  /// | `Unauthorized`           | When an error occurred with the authorization          |
  /// | `ServerError`            | When the status code is between 500 and 599            |
  /// | `ConnectionError`        | When a connection error occurs and cannot be specified |
  /// | `ServerDown`             | When cannot connect to the backend                     |
  /// | `ClientOffline`          | When cannot connect to `yahoo.com`                     |
  /// | `Other`                  | When any other `Exception` occurs                      |
  HttpExceptions type;

  /// The information useful to manage exception.
  Map<String, dynamic> extras;

  /// The response from the call.
  String message;

  /// The response from the call.
  Response? response;

  /// The constructor for the exception.
  ///
  /// Receive the [type] of the exception and [extras] values that can be used to manage the exception.
  HttpCallException(
    HttpExceptions? type, {
    this.message: '',
    this.response,
    Map<String, dynamic>? extras,
  })  : this.type = type ?? HttpExceptions.Other,
        this.extras = extras ?? const {};

  @override
  String toString() {
    switch (this.type) {
      case HttpExceptions.ClientError:
        return 'ClientError';
      case HttpExceptions.ServerError:
        return 'ServerError';
      case HttpExceptions.ConnectionError:
        return 'ConnectionError';
      case HttpExceptions.ServerDown:
        return 'ServerDown';
      case HttpExceptions.ClientOffline:
        return 'ClientOffline';
      case HttpExceptions.ExpiredToken:
        return 'ExpiredToken';
      case HttpExceptions.Unauthorized:
        return 'Unauthorized';
      case HttpExceptions.Other:
        return 'Other';
    }
  }
}
